tutorials/stats/2_services_advanced.py

```diff
@@ -0,0 +1,103 @@
+# %% [markdown]
+"""
+# 2. Services Advanced
+
+The following examples shows how to decorate several functions
+for statistics collection.
+"""
+
+
+# %%
+import os
+import asyncio
+
+from dff.script import Context
+from dff.pipeline import Pipeline, ACTOR, Service, ExtraHandlerRuntimeInfo, to_service
+from dff.stats import StatsStorage, ExtractorPool, StatsRecord
+from dff.stats import default_extractor_pool  # import default pool from addon
+from dff.utils.testing.toy_script import TOY_SCRIPT
+from dff.utils.testing.common import is_interactive_mode
+
+
+# %% [markdown]
+"""
+As is the case with the regular handlers, you can add extractors
+both before and after the target service.
+You can use a handler that runs before the service to compare the pre-service and post-service
+states of the context, measure the running time, etc.
+An example of such handler can be found in the default extractor pool.
+
+Pass before- and after-handlers to the respective parameters of the `to_service` decorator.
+
+As for using multiple pools, you can subscribe your storage to any number of pools.
+
+"""
+
+
+# %%
+extractor_pool = ExtractorPool()
+
+
+@extractor_pool.new_extractor
+async def get_service_state(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
+    # extract execution state of service from info
+    data = {
+        "execution_state": info["component"]["execution_state"],
+    }
+    # return a record to save into connected database
+    return StatsRecord.from_context(ctx, info, data)
+
+
+# %%
+# use together extractor_pool and default_extractor_pool
+# run extract_timing_before before `heavy_service`
+# run get_service_state and extract_timing_after after `heavy_service`
+@to_service(
+    before_handler=[default_extractor_pool["extract_timing_before"]],
+    after_handler=[get_service_state, default_extractor_pool["extract_timing_after"]],
+)
+async def heavy_service(ctx: Context):
+    _ = ctx  # get something from ctx if needed
+    await asyncio.sleep(0.02)
```

Comment:
@RLKRo @pseusys @ruthenian8 
My proposals for  Stats API 
```python
stats_pool = StatsExtractorPool()


@stats_extractor_pool.add_before_handler
async def before_h1(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
    # some code
    return StatsRecord.from_context(ctx, info, data)

@stats_extractor_pool.add_before_handler
async def before_h2(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
    # some code
    return StatsRecord.from_context(ctx, info, data)

@stats_extractor_pool.add_after_handler
async def after_h1(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
    # some code
    return StatsRecord.from_context(ctx, info, data)

@stats_extractor_pool.add_after_handler
async def after_h2(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
    # some code
    return StatsRecord.from_context(ctx, info, data)

@stats_extractor_pool.add_handler
async def h1(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
    # some code
    return StatsRecord.from_context(ctx, info, data)


# %%
# a short example
# @stats_extractor_pool.to_service()
# or
# another short example
# @to_service(stats_extractor_pools = [stats_extractor_pool, default_extractor_pool])
@to_service(
    before_handlers=[h1, ] + stats_extractor_pool.before_handlers,
    after_handlers=[h1,] + stats_extractor_pool.after_handlers,
)
async def heavy_service(ctx: Context):
    _ = ctx  # get something from ctx if needed
    await asyncio.sleep(0.02)
```

Reply:
We can keep previous idea to set explicitly extracting func like `get_service_state`

Reply:
Maybe add the ability to add extractors to custom groups?
```
@extractor_pool.add_extractor("before")
def extract_before(): ...

@extractor_pool.add_extractor("after")
def extract_after(): ...

@to_service(
    before_handler=extractor_pool["before"],
    after_handler=extractor_pool["after"],
)
```

Reply:
Yes, it's a great idea!